dolphin/exception.py

```diff
@@ -22,309 +22,186 @@
-    message = _("User does not have admin privileges.")
+class Invalid(DolphinException):
+    msg_fmt = _("Unacceptable parameters.")
+    code = 400
 
 
-class PolicyNotAuthorized(NotAuthorized):
-    message = _("Policy doesn't allow %(action)s to be performed.")
+class BadRequest(Invalid):
+    msg_fmt = _('The server could not comply with the request since\r\n'
+                'it is either malformed or otherwise incorrect.\r\n')
+    code = 400
 
 
-class Conflict(DolphinException):
-    message = _("%(err)s")
-    code = 409
+class MalformedRequestBody(Invalid):
+    msg_fmt = _("Malformed request body: {0}.")
 
 
-class Invalid(DolphinException):
-    message = _("Unacceptable parameters.")
-    code = 400
+class MalformedRequestUrl(Invalid):
+    msg_fmt = _("Malformed request url.")
 
 
 class InvalidCredential(Invalid):
-    message = _("The credentials are invalid.")
-
-
-class InvalidRequest(Invalid):
-    message = _("The request is invalid.")
+    msg_fmt = _("The credentials are invalid.")
 
 
 class InvalidResults(Invalid):
-    message = _("The results are invalid.")
+    msg_fmt = _("The results are invalid. {0}")
 
 
 class InvalidInput(Invalid):
-    message = _("Invalid input received: %(reason)s.")
+    msg_fmt = _("Invalid input received. {0}")
 
 
 class InvalidName(Invalid):
-    message = _("An invalid 'name' value was provided. %(reason)s")
-
-
-class ValidationError(Invalid):
-    message = "%(detail)s"
+    msg_fmt = _("An invalid 'name' value was provided. {0}")
 
 
 class InvalidContentType(Invalid):
-    message = _("Invalid content type %(content_type)s.")
-
-
-class InvalidHost(Invalid):
-    message = _("Invalid host: %(reason)s")
-
-
-# Cannot be templated as the error syntax varies.
-# msg needs to be constructed when raised.
-class InvalidParameterValue(Invalid):
-    message = _("%(err)s")
-
-
-class InvalidUUID(Invalid):
-    message = _("Expected a uuid but received %(uuid)s.")
-
-
-class InvalidDriverMode(Invalid):
-    message = _("Invalid driver mode: %(driver_mode)s.")
-
-
-class InvalidAPIVersionString(Invalid):
-    message = _("API Version String %(version)s is of invalid format. Must "
-                "be of format MajorNum.MinorNum.")
+    msg_fmt = _("Invalid content type: {0}.")
 
 
-class VersionNotFoundForAPIMethod(Invalid):
-    message = _("API version %(version)s is not supported on this method.")
+class StorageSerialNumberMismatch(Invalid):
+    msg_fmt = _("Storage serial number mismatch. {0}")
 
 
-class InvalidGlobalAPIVersion(Invalid):
-    message = _("Version %(req_ver)s is not supported by the API. Minimum "
-                "is %(min_ver)s and maximum is %(max_ver)s.")
+class StorageAlreadyExists(Invalid):
+    msg_fmt = _("Storage already exists.")
 
 
-class InvalidCapacity(Invalid):
-    message = _("Invalid capacity: %(name)s = %(value)s.")
+class InvalidSNMPConfig(Invalid):
+    msg_fmt = _("Invalid SNMP configuration: {0}")
 
 
 class NotFound(DolphinException):
```

Comment:
Can we rename this class to ResourceNotFound? NotFound is very generic and does not fit.

Reply:
yes, you are right, NotFound is generic which should not be used directly. Actually NotFound is a basic class for any XXXNotFound, it defines the unified status code 404 for all subclasses.

Reply:
But i see this is used directly in some of the code